### PR TITLE
WIP: Issue 337 - Cluster Access API

### PR DESCRIPTION
**Pull request by jofeinstein**
**Changes**
- Cluster access was modeled using an `AllocationUserAttribute` with type `Cluster Access Status`. Now use new model `ClusterAccessRequest` and `ClusterAccessRequestStatusChoice`.
- Added new runners `ClusterAccessRequestCompleteRunner` and `ClusterAccessRequestDenialRunner` to handle completing and denying ClusterAccessRequests
   - Runners use new `EmailStrategies` from #437 
- Added new API endpoint for modifying ClusterAccessRequests at `coldfront/api/allocation/views.py`
- Added command `convert_cluster_attributes_to_cluster_access_request` that converts existing `AllocationUserAttributes` with type `Cluster Account Status` to the new `ClusterAccessRequest` model
- Updated request hub to use the new model
- Updated and added tests
- Changes from #437 merged into this branch
   - failing request hub test was fixed

**Testing**
- Note that secure directory tests are failing on this branch but pass on develop

**TODO**
- [x] Fix comment raised in #337

### CODE REVIEW

**Comment by matthew-li** [comment]
The base branch should be `develop`.

File: coldfront/api/allocation/serializers.py [hunk 1]
```diff
@@ -118,3 +123,78 @@ class Meta:
...
+
+    class Meta:
+        model = ClusterAccessRequest
+        fields = (
+            'id', 'status', 'completion_time', 'cluster_uid',
+            'username', 'billing_activity', 'allocation_user')
+        extra_kwargs = {
+            'id': {'read_only': True},
+            'completion_time': {'required': False, 'allow_null': True},
+            'allocation_user': {'required': True,
```
Comment (matthew-li) [inline coldfront/api/allocation/serializers.py:154]: Is this already covered by L139?

File: coldfront/api/allocation/serializers.py [hunk 2]
```diff
@@ -118,3 +123,78 @@ class Meta:
...
+                                'allow_null': False,
+                                'read_only': True}
+        }
+
+    def validate(self, data):
+        # If the status is being changed to 'Complete', ensure that a
```
Comment (matthew-li) [inline coldfront/api/allocation/serializers.py:160]: The status in question should be 'Active'.

File: coldfront/api/allocation/serializers.py [hunk 3]
```diff
@@ -118,3 +123,78 @@ class Meta:
...
+        # completion_time, username, and cluster_uid are given.
+        if 'status' in data and data['status'].name == 'Active':
+            messages = []
+            for field in ['completion_time', 'username', 'cluster_uid']:
```
Comment (matthew-li) [inline coldfront/api/allocation/serializers.py:164]: Is the `for` loop needed? Can we just check each?

File: coldfront/api/allocation/tests/test_allocation_base.py [hunk 4]
```diff
@@ -21,6 +23,19 @@ def setUp(self):
         """Set up test data."""
         super().setUp()
 
+        # Create default choices.
+        call_command('add_default_user_choices')
```
Comment (matthew-li) [inline coldfront/api/allocation/tests/test_allocation_base.py:27]: The second command should already be called by a parent class (`TestAPIBase`). Could the first command be placed there too, and these two removed from here, without issue?

File: coldfront/api/allocation/views.py [hunk 5]
```diff
@@ -112,3 +140,49 @@ def get_queryset(self):
...
+                # Pop read only fields
+                cluster_uid = serializer.validated_data.pop('cluster_uid', None)
+                username = serializer.validated_data.pop('username', None)
+                serializer.validated_data.pop('allocation_user', None)
+
+                instance = serializer.save()
+
+                if instance.status.name == 'Active':
+                    runner = ClusterAccessRequestCompleteRunner(instance)
+                    runner.run(cluster_uid=cluster_uid,
```
Comment (matthew-li) [inline coldfront/api/allocation/views.py:171]: I imagine this works as is, but the parameters are not named, and the arguments are in the wrong order.

File: coldfront/api/allocation/views.py [hunk 6]
```diff
@@ -112,3 +140,49 @@ def get_queryset(self):
...
+                    runner.run()
+        except Exception as e:
+            message = f'Rolling back failed transaction. Details:\n{e}'
+            logger.exception(message)
+            raise APIException('Internal server error.')
+
+    @swagger_auto_schema(
+        manual_parameters=[authorization_parameter],
+        operation_description=(
+                'Updates one or more fields of the ProjectUserRemovalRequest '
```
Comment (matthew-li) [inline coldfront/api/allocation/views.py:184]: Wrong model.

**Review by matthew-li — CHANGES_REQUESTED** [review]
Thanks for making these changes--I've left some minor comments, but otherwise, the changes look reasonable.

I'll run the test suites and do manual testing afterward.

File: coldfront/core/allocation/management/commands/convert_cluster_attributes_to_cluster_access_request.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
+import logging
+
+from django.core.management import BaseCommand
+
+from coldfront.core.allocation.models import AllocationAttributeType, \
+    AllocationUserAttribute, ClusterAccessRequestStatusChoice, \
+    ClusterAccessRequest
+
+
+"""An admin command that approves AllocationRenewalRequests for a
```
Comment (matthew-li) [inline coldfront/core/allocation/management/commands/convert_cluster_attributes_to_cluster_access_request.py:10]: Copied-and-pasted docstring.

**Comment by matthew-li** [comment]
@jofeinstein I added a `history` field to `ClusterAccessRequests` in 10cc3b4.

Additionally, I rewrote the conversion command in 87b8674 to:
- Take the date as an input (LRC's is different.)
- Derive request information from the `history` of each `AllocationUserAttribute`. This helps to:
    - Accurately determine when the request was made and completed (it seems like `created` isn't necessarily the most accurate),
    - Allow multiple request flows to be detected and turned into requests (i.e., get added, get removed, then get added again) (although this case probably hasn't arisen),
    - Protect against the edge case wherein an attribute was set to "Denied" because of the user being removed from the project (which we don't want to record as a request), and
    - Allow it to be run multiple times without creating duplicates (nice to have).

Also, we can always reconstruct these requests from the `history`, so if this algorithm isn't perfect (ideally, it would create `HistoricalClusterAccessRequests` with the proper `history_dates`, but that got complicated to implement), it can be improved in the future.

Can you review the changes and test the command? You'll need to un-migrate back to `0010` and then re-migrate (and run `add_allocation_defaults`) first.

**Comment by jofeinstein** [comment]
The changes make sense and the command still runs. I've never done anything with the history so that was pretty neat.

**Comment by matthew-li** [comment]
@jofeinstein Thanks for checking. There are some failing tests in `coldfront.api.allocation.tests.allocations.test_cluster_access_request` as a result of the change from "Active" to "Complete", (among other issues, maybe) but other than that, tests are passing.

**Comment by jofeinstein** [comment]
Test suite passes on my local VM.

**Pull request merged**